File: JavaExercise/src/top/interview/MaxProfit.java
package top.interview;

/***
 * https://leetcode.com/explore/featured/card/top-interview-questions-easy/97/dynamic-programming/572/
 *
 * 找到最佳买卖股票的十几人，只允许交易一次
 */
public class MaxProfit {
    public MaxProfit(){

    }
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0){
            return 0;
        } else{
            int maxProfit = 0;
            int min = prices[0];
            int n = prices.length;
            for (int i = 1; i < n; i++){
                maxProfit = Math.max(maxProfit, prices[i] - min);
                if (prices[i] < min){
                    min = prices[i];
                }
            }
            return maxProfit;
        }
    }

    public static void main(String[] args) {
        int[] inout = new int[]{7,1,5,3,6,4};
        System.out.println(new MaxProfit().maxProfit(inout));
    }
}

File: top1_interview/max_profit.py
# -*- coding: utf-8 -*-
"""

   File Name：     max_profit
   Author :       jing
   Date：          2019-12-08
   https://leetcode.com/explore/featured/card/top-interview-questions-easy/92/array/564/
   找到所有连续递增的子序列
"""


class Solution:
    def maxProfit(self, prices) -> int:
        profit = 0
        submin = 0
        for i in range(1, len(prices)):
            if prices[i-1] <= prices[i]:
                if i == len(prices) - 1:
                    profit = profit + (prices[i] - prices[submin])
                continue
            else:
                 profit = profit + (prices[i-1] - prices[submin])
                 submin = i
        return profit


if __name__ == '__main__':
    print(Solution().maxProfit([1, 5, 5, 6, 4, 10]))
